### PR TITLE
[Impeller] Polyline generation optimization.

### PR DESCRIPTION
**Pull request by dnfield**
The proposed approach has several advantages:

- It is _much_ faster for quadratics, since we no longer have to
  elevate them to cubics.
- It is slightly slower for cubics, but also produces fewer and
  more accurate points on average, which results in faster
  tessellation and a net decrease in CPU time.
- The core of the algorithm is parallelizable, which should make it
  easier to use on the GPU without significant changes.

Before this change:

```
2022-10-14T11:11:14-07:00
Running out/host_release/geometry_benchmarks
Run on (10 X 2400 MHz CPU s)
CPU Caches:
  L1 Data 64 KiB (x10)
  L1 Instruction 128 KiB (x10)
  L2 Unified 4096 KiB (x5)
Load Average: 7.00, 5.70, 5.09
------------------------------------------------------------------------------------------
Benchmark                                Time             CPU   Iterations UserCounters...
------------------------------------------------------------------------------------------
BM_Polyline/cubic_polyline           22642 ns        22585 ns        31151 SinglePointCount=251 TotalPointCount=7.8189M
BM_Polyline/cubic_polyline_tess      78041 ns        77847 ns         9114 SinglePointCount=251 TotalPointCount=2.28761M
BM_Polyline/quad_polyline            22622 ns        22562 ns        31047 SinglePointCount=236 TotalPointCount=7.32709M
BM_Polyline/quad_polyline_tess       74762 ns        74563 ns         9318 SinglePointCount=236 TotalPointCount=2.19905M
```

After:

```
2022-10-14T11:12:26-07:00
Running out/host_release/geometry_benchmarks
Run on (10 X 2400 MHz CPU s)
CPU Caches:
  L1 Data 64 KiB (x10)
  L1 Instruction 128 KiB (x10)
  L2 Unified 4096 KiB (x5)
Load Average: 4.32, 5.21, 4.96
------------------------------------------------------------------------------------------
Benchmark                                Time             CPU   Iterations UserCounters...
------------------------------------------------------------------------------------------
BM_Polyline/cubic_polyline           35580 ns        35574 ns        19980 SinglePointCount=155 TotalPointCount=3.0969M
BM_Polyline/cubic_polyline_tess      70250 ns        70240 ns         9998 SinglePointCount=155 TotalPointCount=1.54969M
BM_Polyline/quad_polyline            15065 ns        15062 ns        46802 SinglePointCount=144 TotalPointCount=6.73949M
BM_Polyline/quad_polyline_tess       48014 ns        47992 ns        14644 SinglePointCount=144 TotalPointCount=2.10874M
```

I want the benchmark to land first.

Bottom line is that creating polylines for quadratics gets much faster/leaner/more accurate. Creating polylines for cubics gets slightly slower but creates fewer more accurate points. This is a big win for quadratics, and a small win for cubics once you include tessellation.

The other nice property is that these algorithms are more translatable to shaders than the recursive algorithm.

### CODE REVIEW

**Comment by dnfield** [comment]
Argh. One of these commits had a ncie message I'll try to lift into the body of the PR...

**Comment by dnfield** [comment]
This is ready for review.

File: impeller/tessellator/c/tessellator.h [hunk 1]
```diff
@@ -42,7 +42,8 @@ IMPELLER_API void Close(PathBuilder* builder);
 
 IMPELLER_API struct Vertices* Tessellate(PathBuilder* builder,
                                          int fill_type,
-                                         Scalar scale,
+                                         Scalar tolerance,
+                                         // Unused, remaining for ABI stability.
```
Comment (chinmaygarde) [inline impeller/tessellator/c/tessellator.h:46]: Can we clean this up at all?

Comment (dnfield) [reply impeller/tessellator/c/tessellator.h:46]: vector_graphics is potentially using these parameters and would crash without them. Although I don't think we actually have anything using them right now - I'll just drop them for now and we can clean up vg separately.

File: impeller/geometry/path_component.h [hunk 2]
```diff
@@ -117,15 +99,26 @@ struct CubicPathComponent {
...
+  // generates a polyline from those quadratics.
+  //
+  // See the note on QuadraticPathComponent::CreatePolyline for references.
   std::vector<Point> CreatePolyline(
-      const SmoothingApproximation& approximation) const;
+      Scalar tolerance = kDefaultCurveTolerance) const;
 
   std::vector<Point> Extrema() const;
 
+  std::vector<QuadraticPathComponent> ToQuadratics(Scalar accuracy) const;
```
[the remark this answers is not in the document]
Comment (dnfield) [reply impeller/geometry/path_component.h:111]: done, but with the `s` still :)

File: impeller/geometry/path_component.cc [hunk 3]
```diff
@@ -85,10 +81,57 @@ Point QuadraticPathComponent::SolveDerivative(Scalar time) const {
...
+  auto dd = d01 - d12;
+  auto cross = (p2 - p1).Cross(dd);
+  auto x0 = d01.Dot(dd) * 1 / cross;
+  auto x2 = d12.Dot(dd) * 1 / cross;
+  auto scale = abs(cross / (hypot(dd.x, dd.y) * (x2 - x0)));
+
+  auto a0 = ApproximateParabolaIntegral(x0);
+  auto a2 = ApproximateParabolaIntegral(x2);
+  Scalar val = 0.f;
+  if (isfinite(scale)) {
```
[the remark this answers is not in the document]
Comment (dnfield) [reply impeller/geometry/path_component.cc:111]: Done

**Comment by dnfield** [comment]
https://flutter-engine-perf.skia.org/e/?keys=X18cd97edd746da8f4ac526310da80177
